generator/srtm_parser.cpp

```diff
@@ -0,0 +1,172 @@
+};
+}  // namespace
+
+// SrtmTile ----------------------------------------------------------------------------------------
+SrtmTile::SrtmTile()
+{
+  Invalidate();
+}
+
+SrtmTile::SrtmTile(SrtmTile && rhs) : m_data(move(rhs.m_data)), m_valid(rhs.m_valid)
+{
+  rhs.Invalidate();
+}
+
+void SrtmTile::Init(string const & dir, ms::LatLon const & coord)
+{
+  Invalidate();
+
+  string const base = GetBase(coord);
+  string const cont = dir + base + ".SRTMGL1.hgt.zip";
+  string file = base + ".hgt";
+
+  UnzipMemDelegate delegate(m_data);
+  try
+  {
+    ZipFileReader::UnzipFile(cont, file, delegate);
+  }
+  catch (ZipFileReader::LocateZipException e)
+  {
+    // Sometimes packed file has different name. See N39E051 measure.
+    file = base + ".SRTMGL1.hgt";
+
+    ZipFileReader::UnzipFile(cont, file, delegate);
+  }
+
+  if (!delegate.m_completed)
+  {
+    LOG(LWARNING, ("Can't decompress SRTM file:", cont));
+    Invalidate();
+    return;
+  }
+
+  if (m_data.size() != kSrtmTileSize)
+  {
+    LOG(LWARNING, ("Bad decompressed SRTM file size:", cont, m_data.size()));
+    Invalidate();
+    return;
+  }
+
+  m_valid = true;
+}
+
+SrtmTile::THeight SrtmTile::GetHeight(ms::LatLon const & coord)
+{
+  if (!IsValid())
+    return kInvalidHeight;
+
+  double ln = coord.lon - static_cast<int>(coord.lon);
+  if (ln < 0)
+    ln += 1;
+  double lt = coord.lat - static_cast<int>(coord.lat);
+  if (lt < 0)
+    lt += 1;
+  lt = 1 - lt;  // from North to South
+
+  size_t const row = kArcSecondsInDegree * lt;
+  size_t const col = kArcSecondsInDegree * ln;
+
+  size_t const ix = row * (kArcSecondsInDegree + 1) + col;
+
+  if (ix >= Size())
+    return kInvalidHeight;
+  return ReverseByteOrder(Data()[ix]);
+}
+
+string SrtmTile::GetBase(ms::LatLon coord)
+{
+  ostringstream ss;
+  if (coord.lat < 0)
+  {
+    ss << "S";
+    coord.lat *= -1;
+    coord.lat += 1;
+  }
+  else
+  {
+    ss << "N";
+  }
+  ss << setw(2) << setfill('0') << static_cast<int>(coord.lat);
+
+  if (coord.lon < 0)
+  {
+    ss << "W";
+    coord.lon *= -1;
+    coord.lon += 1;
+  }
+  else
+  {
+    ss << "E";
+  }
+  ss << setw(3) << static_cast<int>(coord.lon);
+  return ss.str();
+}
+
+void SrtmTile::Invalidate()
+{
+  m_data.clear();
+  m_data.shrink_to_fit();
+  m_valid = false;
+}
+
+// SrtmTileManager ---------------------------------------------------------------------------------
+SrtmTileManager::SrtmTileManager(string const & dir) : m_dir(dir) {}
+
+SrtmTile::THeight SrtmTileManager::GetHeight(ms::LatLon const & coord)
+{
+  string const base = SrtmTile::GetBase(coord);
+  auto it = m_tiles.find(base);
+  if (it == m_tiles.end())
+  {
+    SrtmTile tile;
+    try
+    {
+      tile.Init(m_dir, coord);
+    }
+    catch (RootException const & e)
+    {
+      LOG(LWARNING, ("Can't init SRTM tile:", base, "reason:", e.Msg()));
+    }
+
+    // It's OK to store even invalid files and return invalid height
+    // for them later.
+    m_tiles.emplace(base, move(tile));
```

Comment:
it = m_tiles.emplace(base, move(tile)).first;
...
return it->second.GetHeight(coord);
?